File: leetcode/c++/2134_minimum_swaps_to_group_all_1s_together-_ii.cpp
// https://leetcode.cn/problems/minimum-swaps-to-group-all-1s-together-ii/description/

#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int minSwaps(vector<int>& nums) {
        int k = 0, res = 1e5 + 10, n = nums.size();
        vector<int> array;
        for (int i = 0; i < 2 * n; i ++){
            if (i < n)
                k += nums[i];
            array.push_back(nums[i % n]);
        }          
        for (int l = 0, r = 0, tmp = 0; r < 2*n; r++) {
            tmp += abs(array[r] - 1);
            if (r >= l + k) {
                tmp -= abs(array[l++] - 1);
            }
            if (r == l + k -1)
                res = min(res, tmp);
        }
        return res;
    }
};

// 时间复杂度O(n), 空间复杂度O(n)

File: leetcode/golang/2134_minimum_swaps_to_group_all_1s_together-_ii.go
func abs(num int) int{
    if num < 0 {
        return -num
    }
    return num
}

func minSwaps(nums []int) int {
    k := 0
    n := len(nums)
    res := n
    array := make([]int, 0, 2 * n) 
    for i := 0; i < 2 * n; i++ {
        if i < n {
            k += nums[i]
        }
        array = append(array, nums[i % n])
    }
    for l , r, tmp := 0, 0, 0; r < 2 * n; r++ {
        tmp += abs(array[r] - 1)
        if r >= l + k {
            tmp -= abs(array[l] - 1)
            l++
        }
        if r == l + k - 1 {
            res = min(res, tmp)
        }
    }
    return res
}